rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,165 @@
+---
+authors: Ryan Clark (ryan.clark@goteleport.com)
+state: draft
+---
+
+# RFD 85 - Better React testing, remove default exports
+
+## What
+
+Improve the webapps codebase by changing the way we test our components and write our component stories.
+
+## Why
+
+The current convention is to use Storybook to render our large, full components in their different states.  This has a
+couple of knock-on effects.
+
+### Component Architecture
+
+As it's hard to mock data in Storybook, we end up exporting a component that is suitable for rendering in Storybook,
+such as:
+
+```typescript jsx
+export function DesktopSession(props: State) {
+  const {
+    directorySharingState,
+    setDirectorySharingState,
+    clipboardState,
+    fetchAttempt,
+    tdpConnection,
+    disconnected,
+    wsConnection,
+    setTdpConnection,
+  } = props;
+  // etc
+
+  return (
+    <Flex flexDirection="column">
+      // etc
+    </Flex>
+  );
+}
+```
+
+This component takes all of its state and business logic from the props passed in.
+
+For use in the actual application, we wrap the component above in a default export which populates the props the
+component needs by using a hook.
+
+```typescript jsx
+export default function Container() {
+  const state = useDesktopSession();
+  return <DesktopSession {...state} />;
+}
+```
+
+
+### Testing
+
+As a result of the above, we end up testing the `DesktopSession` component instead of the actual component (`Container`) that is
+rendered in the UI.
+
+A story may look like this:
+
+```typescript jsx
+export const FetchError = () => (
+  <DesktopSession
+    {...props}
+    fetchAttempt={{ status: 'failed', statusText: 'some fetch  error' }}
+    tdpConnection={{ status: 'success' }}
+    wsConnection={'open'}
+    disconnected={false}
+  />
+);
+```
+
+Which is then tested like so:
+
+```typescript jsx
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the UI part of the application,
+but skipping all of the business logic.
+
+Storybook isn't best suited for the way we are using it. As a result of testing our large components primarily via stories, we have
+ended up only testing our UI and nothing else.
+
+## Details
+
+### Changing how we write stories
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components. We should therefore
+restrict usage of Storybook mostly to our shared component library, and test our fullscreen components another way.
+
+It's still okay to create our larger components in Storybook (if the developer wants to) but we should not be using these stories in our Jest tests.
+
+It might be worthwhile looking into something like [Chromatic](https://www.chromatic.com/) for visual UI testing.
```

Comment:
```suggestion
To get a true visual test of the UI while eliminating most of the downsides of snapshot tests, we will switch over to using [Chromatic](https://www.chromatic.com/).
```

Reply:
I spoke with @zmb3 and he said to leave Chromatic out of the scope of this RFD, but he'll take a look at Chromatic and see if we can kick things off

Reply:
As it's not in the scope of this PR maybe add a few other alternatives here like [playwrite](https://playwright.dev/) and [puppeteer](https://pptr.dev/).

Also why is it worthwhile, what does using a tool like this give us that using snapshots doesn't?

This content of this section doesn't actually answer the content of the header "How will we change the UI". Maybe this is just a case of changing the hierarchy of this heading to be higher.